pkg/monitortests/kubeapiserver/crdversionchecker/monitortest.go

```diff
@@ -0,0 +1,486 @@
+		return summary.ChangedCRDs[i].Name < summary.ChangedCRDs[j].Name
+	})
+
+	return summary
+}
+
+// computeCRDChanges determines what changed between two versions of a CRD.
+func computeCRDChanges(name string, before, after CRDInfo) *CRDChangeSummary {
+	beforeVersions := make(map[string]bool)
+	afterVersions := make(map[string]bool)
+	var beforeStorage, afterStorage string
+
+	for _, v := range before.Versions {
+		beforeVersions[v.Name] = true
+		if v.Storage {
+			beforeStorage = v.Name
+		}
+	}
+
+	for _, v := range after.Versions {
+		afterVersions[v.Name] = true
+		if v.Storage {
+			afterStorage = v.Name
+		}
+	}
+
+	// Find added versions
+	var addedVersions []string
+	for v := range afterVersions {
+		if !beforeVersions[v] {
+			addedVersions = append(addedVersions, v)
+		}
+	}
+	sort.Strings(addedVersions)
+
+	// Find removed versions
+	var removedVersions []string
+	for v := range beforeVersions {
+		if !afterVersions[v] {
+			removedVersions = append(removedVersions, v)
+		}
+	}
+	sort.Strings(removedVersions)
+
+	// Check if storage version changed
+	storageChanged := beforeStorage != afterStorage
+
+	// Only return a change summary if something actually changed
+	if len(addedVersions) == 0 && len(removedVersions) == 0 && !storageChanged {
+		return nil
+	}
+
+	return &CRDChangeSummary{
+		Name:            name,
+		AddedVersions:   addedVersions,
+		RemovedVersions: removedVersions,
+		StorageChanged:  storageChanged,
+		OldStorage:      beforeStorage,
+		NewStorage:      afterStorage,
+	}
+}
+
+// =============================================================================
+// Validation Checks
+// =============================================================================
+
+// checkNewVersionsNotStoredImmediately verifies that newly added API versions
+// are not immediately set as the storage version.
+//
+// Rationale: When a new API version is introduced during an upgrade, existing
+// data in etcd is still stored in the old format. Setting the new version as
+// the storage version immediately would require a migration. The safe approach
+// is to serve both versions but keep the old version as storage until all
+// existing objects have been migrated.
+//
+// This check FAILS if:
+// - A CRD has a new version added
+// - The new version is marked as the storage version
+// - The old version is no longer the storage version
+//
+// This check PASSES if:
+// - New versions are added but the old storage version is preserved
+// - Both old and new versions are served
+func (c *crdVersionChecker) checkNewVersionsNotStoredImmediately() []*junitapi.JUnitTestCase {
+	const testName = "[sig-api-machinery] CRDs with new API versions should not change storage version immediately"
+
+	if c.beforeSnapshot == nil || c.afterSnapshot == nil {
+		return []*junitapi.JUnitTestCase{{
+			Name: testName,
+			SkipMessage: &junitapi.SkipMessage{
+				Message: "Missing CRD snapshots, cannot perform check",
```

Comment:
We alway return a skip when not running during an upgrade? Is that what we want vs just outputting an empty JUnit? Not sure what the correct etiquette would be here